python/tidy.py

```diff
@@ -159,45 +159,52 @@ def stdout_redirect(where):
 
 
 def check_lock(file_name, contents):
+    def find_reverse_dependencies(dependency, version, content):
+        dependency_prefix = "{} {}".format(dependency, version)
+        for package in itertools.chain([content["root"]], content["package"]):
+            for dependency in package.get("dependencies", []):
+                if dependency.startswith(dependency_prefix):
+                    yield package["name"]
+
     if not file_name.endswith(".lock"):
         raise StopIteration
-    contents = contents.splitlines(True)
-    idx = 1
-    packages = {}
 
     # package names to be neglected (as named by cargo)
-    exceptions = ["libc"]
-
-    while idx < len(contents):
-        content = contents[idx].strip()
-        if 'name' in content:
-            base_name = content.split('"')[1]
-            # we need the base package because some other package might demand a new version in the following lines
-            packages[base_name] = contents[idx + 1].split('"')[1], idx + 2, base_name
-        if 'dependencies' in content:
-            idx += 1
-            while contents[idx].strip() != ']':
-                package = contents[idx].strip().strip('",').split()
-                name, version = package[0], package[1]
-                if name not in packages:    # store the line number & base package name for future comparison
-                    packages[name] = (version, idx + 1, base_name)
-                elif all([packages[name][0] != version, name not in exceptions, base_name not in exceptions]):
-                    line = idx + 1
-                    version_1 = tuple(map(maybe_int, packages[name][0].split('.')))
-                    version_2 = tuple(map(maybe_int, version.split('.')))
-                    if version_1 < version_2:       # get the line & base package containing the older version
-                        packages[name], (version, line, base_name) = (version, line, base_name), packages[name]
-
-                    message = 'duplicate versions for package "%s"' % name
-                    error = '\n\t\033[93mexpected maximum version "{}"\033[0m'.format(packages[name][0]) + \
-                            '\n\t\033[91mbut, "{}" demands "{}"\033[0m' \
-                            .format(base_name, version)
-                    suggest = "\n\t\033[93mtry upgrading with\033[0m " + \
-                              "\033[96m./mach cargo-update -p {}:{}\033[0m" \
-                              .format(name, version)
-                    yield (line, message + error + suggest)
-                idx += 1
-        idx += 1
+    exceptions = ["libc", "byteorder", "cocoa"]
+
+    import toml
+    content = toml.loads(contents)
+
+    packages = {}
+    for package in content.get("package", []):
```

Comment:
Nit: Shall we have the import block at the top along with the others? Also, separate the declarations from the iterations? (for the sake of beauty! :P)